HLA-1342: Documentation for S_REGION code change (#1891)
spacetelescope/drizzlepac: doc/source/mast_data_products/svm_processing.rst

just the rectangular shape of the final drizzle array. This allows the archive
to provide a preview of the drizzle product footprints in their all-sky map to
assist users in selecting the data most suited for their search.

The computation of this keyword relies on creating a filled shape, finding the border
pixels of the shape, ordering the vertices in a counter-clockwise manner, and
simplifying the shape by removing unnecessary vertices. The resulting vertices
are then converted to sky coordinates (degrees). Our strategy relies on two
complementary techniques named dilation and erosion, which add and remove pixels,
respectively.

In the code, the top level call functions are in :py:mod:`~drizzlepac.haputils.processing_utils`
where the compute_sregion() function calls the find_footprint function.
This function makes calls to two methods in the SkyFootprint class in :py:mod:`~drizzlepac.haputils.cell_utils`
that build up a total mask (extract_mask), and find the vertices/corners of that mask
(find_corners).

Dilation and Erosion
********************
We use the mathematical morphology operations of binary erosion and binary dilation

> in different orders for two purposes. Using erosion and then dilation removes pixels
> and then inflates the outer region. This has the effect of isolating island pixels and
> removing noise. Doing the opposite, erosion and then dilation, will result in more-unique
> peninsulas and edge features being saved but smoothed to some degree.

Reviewer: ```suggestion
in different orders for two purposes. Using erosion and then dilation (the morphological
"opening" operation) removes isolated pixels and narrow tendrils.  
This has the effect of isolating islands of pixels that are thinly connected to one another and
removing noise. Doing the opposite, dilation and then erosion (the morphological "closing"
operation), fills in small holes and gaps to join islands and smooth the outer edge
to some degree.
```

Extracting a Mask
*****************
Extracting the the mask first involves using a ndimage.binary_fill_holes to fill
in any holes surrounded by data. The code then uses dilation to keep edge features
while smoothing the outer edge. Erosion is then used to get a mask that has a similar
shape and size to the original mask. We then combine the resulting mask with the original

> mask to ensure we retain edge features while yielding a smoother edge.

Reviewer: ```suggestion
mask to ensure we retain features that touch the image edges.
```

Finding Border Pixels
*********************
The code traces the edge of the footprint to isolate only the edge pixels. It
then uses erosion and then dilation to remove noise and island pixels. It then
applies a gaussian filter to the edge pixels to ultimately create a smoother S_REGION.

> The resulting edge pixels are then simplified.

Reviewer: ```suggestion
To find all the edge pixels, an eroded version of the mask is subtracted from the mask,
which leaves behind only the pixels around the edge.  A box search algorithm is used to 
connect the edge pixels in a path going counter-clockwise around the image edge.
Finally, the resulting edge path is simplified.
```

Simplifying the Border
**********************
In order to decrease the number of vertices for our S_REGION we use the simplify-polyline
package. A call to simplify-polyline.simplify looks at each set of three
consecutive vertices and determines if the errors in the line changes significantly
when the middle pixel is removed. If it is below an set error (min_dist), the code

> removes the pixel. It should be noted that these three pixels can be far apart.

Reviewer: ```suggestion
removes the pixel. This algorithm is repeated until no more pixels are found to remove.
It should be noted that in later iterations, these three pixels can be far apart.
```
